### ISSUE TITLE
Permit multiline in URL

### ISSUE THREAD

**Issue by am11** [sass/libsass#1096]
Test:

```scss
@import url("a\
b");
```

LibSass master (0dd65439366bd4b6d9fce42a7f6a45990df5e836) errors:

> URI is missing ')'

LibSass 3.1 outputs:

```css
@import url("a\
b");
```

RubySass v3.4.12 outputs:

```css
@import url("ab");
```

---

Related test:

```scss
// without quotes
@import url(a\
b);
```

RubySass throws the following error:

> Invalid CSS after "`@import` url(a": expected ")", was "\"

LibSass master (0dd65439366bd4b6d9fce42a7f6a45990df5e836) errors:

> URI is missing ')'

LibSass 3.1 outputs:

```css
@import url(a\
b);
```

**Comment by xzyfer** [comment]
Specs added https://github.com/sass/sass-spec/pull/345

**Comment by am11** [comment]
@mgreter, there is still one crazy case which is inconsistent (perhaps not worth fixing?).

(added periods to highlight spaces)

```scss
@import url("a.. .\.. .
 . . b");
```

Ruby Sass:

```css
@import url("a.. ... .\a  . . b");
```

Current libsass master (28ee0884243820a34c3458e26b36fc59ccefd9b3):

```css
@import url("a.. ... .
\a  . . b");
```

Not sure about logic behind ruby-sass implementation. I was expecting something like this:

```css
@import url("a.. ... .\  . . b");
```

**Comment by xzyfer** [comment]
This still fails the spec.

**Comment by xzyfer** [comment]
Apologies, it looks like my local sass-spec is having issues.

**Comment by xzyfer** [comment]
This fix was a false positive due to https://github.com/sass/sass-spec/pull/380. The spec still fails.

**Comment by xzyfer** [comment]
Related https://github.com/sass/libsass/issues/1294

**Comment by am11** [comment]
@xzyfer,

quickly tested current master with the following input:

```scss
/* url-test1: without quotes */
@import url(a
b);

/* url-test1: with quotes */
@import url("a
b");

/* url-test2: with quotes */
@import url("a   \d \\d \\\d \\\\d
    c \t \n \\t \\n \\\t \\\n \\\\t \\\\n");
  
/* url-test2: without quotes */
@import url(a   \d \\d \\\d \\\\d
    c \t \n \\t \\n \\\t \\\n \\\\t \\\\n);
```

LibSass:

```css
/* url-test1: without quotes */
@import url(a b);
/* url-test1: with quotes */
\a b"); url("a
/* url-test2: with quotes */
\a     c t n \\t \\n \\t \\n \\\\t \\\\n");
/* url-test2: without quotes */
@import url(a \d \\d \\\d \\\\d c \t \n \\t \\n \\\t \\\n \\\\t \\\\n);
```

RubySass:

```css
/* url-test1: without quotes */
@import url(a b);
/* url-test1: with quotes */
@import url("a\a b");
/* url-test2: with quotes */
@import url("a   
\\d \\
\\\\d\a     c t n \\t \\n \\t \\n \\\\t \\\\n");
/* url-test2: without quotes */
@import url(a \d \\d \\\d \\\\d c \t \n \\t \\n \\\t \\\n \\\\t \\\\n);
```

IMO,  `url-test1: with quotes` is more important than rest of them.

**Comment by xzyfer** [comment]
Interesting. I just tested with 0f929a5 and got identical output (minus deprecation warnings)

```
/* url-test1: without quotes */
@import url(a b);
/* url-test1: with quotes */
@import url("a\a b");
/* url-test2: with quotes */
\\\\d\a     c t n \\t \\n \\t \\n \\\\t \\\\n");
/* url-test2: without quotes */
@import url(a \d \\d \\\d \\\\d c \t \n \\t \\n \\\t \\\n \\\\t \\\\n);
```

**Comment by am11** [comment]
Yup, I tested on both Mac and Windows, make result is identical to yours. `url-test1` in Windows is inverted from the middle. `url-test2` in both are differently incorrect. I used libsass a508a7c and sassc https://github.com/sass/sassc/commit/94838a90c403120af9da5f71a8daf79cf3ec82fe:

```sh
~/Documents/projects/libsass/sassc/bin/sassc ~/Documents/temp/blah.scss
```

**Comment by xzyfer** [comment]
@am11 sorry I'm not following your previous comment. Please be more explicitly.

Can you please post the output you're seeing on Windows, and how it differs from Mac.

Interestingly I'm unable to get the Ruby Sass output you've provided locally (on Mac), but I do see it on sassmeister.

**Comment by am11** [comment]
I am running Mac and Windows 10 via Parallels.

Mac:

```sh
~/Documents/projects/libsass/sassc/bin/sassc ~/Documents/temp/blah.scss
```

```css
/* url-test1: without quotes */
@import url(a b);
/* url-test1: with quotes */
@import url("a\a b");
/* url-test2: with quotes */
\\\\d\a     c t n \\t \\n \\t \\n \\\\t \\\\n");
/* url-test2: without quotes */
@import url(a \d \\d \\\d \\\\d c \t \n \\t \\n \\\t \\\n \\\\t \\\\n);
```

Windows:

```ps
~/Documents/projects/libsass/sassc/bin/sassc ~/Documents/temp/blah.scss
```

```css
/* url-test1: without quotes */
@import url(a b);
/* url-test1: with quotes */
\a b"); url("a
/* url-test2: with quotes */
\a     c t n \\t \\n \\t \\n \\\\t \\\\n");
/* url-test2: without quotes */
@import url(a \d \\d \\\d \\\\d c \t \n \\t \\n \\\t \\\n \\\\t \\\\n);
```

`/* url-test1: with quotes */` on Windows is inverted from the middle. 

`/* url-test2: with quotes */` on Mac and Windows are incorrect but different.

**Comment by am11** [comment]
@xzyfer, sorry I was looking at the CLI output which was broken in both PowerShell and Terminal in different ways.

I just piped to a file and saw the output in editor.

```sh
scssc  blah.scss > blah.css
```

Mac (TextEdit)

```css
/* url-test1: without quotes */
@import url(a b);
/* url-test1: with quotes */
@import url("a\a b");
/* url-test2: with quotes */
@import url("a   
\\d \\
\\\\d\a     c t n \\t \\n \\t \\n \\\\t \\\\n");
/* url-test2: without quotes */
@import url(a \d \\d \\\d \\\\d c \t \n \\t \\n \\\t \\\n \\\\t \\\\n);
```

Windows (Visual Studio and then HexEdit to check what's going on):

```css
/* url-test1: without quotes */
@import url(a b);
/* url-test1: with quotes */
@import url("a
\a b");
/* url-test2: with quotes */
@import url("a   
\\d \\
\\\\d
\a     c t n \\t \\n \\t \\n \\\\t \\\\n");
/* url-test2: without quotes */
@import url(a \d \\d \\\d \\\\d c \t \n \\t \\n \\\t \\\n \\\\t \\\\n);
```

Note that in Windows output, the extra line before each `\a` in `\r` (HEX `0D`). I haven't tested whether it is the raw result of compiler or it is shell redirection inserting it there. Whichever it is, this issue is fixed. Thank you! :trophy: :+1:

**Comment by am11** [comment]
@xzyfer, I have confirmed that `\r` (hex: `0D`) is added before both `\a` occurrences, by placing `printf("size >>> %lu\n", strlen(output_string));` before: https://github.com/sass/sassc/blob/94838a9/sassc.c#L46.

When compiling [this input sample](https://github.com/sass/libsass/issues/1096#issuecomment-167531680) on Mac it prints: `size >>> 308` while on Windows: `size >>> 310`.

Any ideas where this `\r` it is coming from?

**Comment by xzyfer** [comment]
@am11 nothing comes to mind. Also I don't think this patch affected quoted
imports. Can you confirm whether this issue existed before this patch?

**Comment by xzyfer** [comment]
Also, if you include debugger.cpp in context. cpp and add
`debug_ast(root);` on line 657 it'll dump the ast tree before the output
phase.

**Comment by xzyfer** [comment]
Also if you include `debugger.cpp` in `context.cpp` and add

```
debug_ast(root);
```
on line 657 the AST tree will be dumped to the console just before the output phase begins.

**Comment by am11** [comment]
With debugger, on Mac:

```
####################################################################
  Block 0x7fddcbc1d410 (0@[0:0]-[16:0]) [root] 0
  Comment 0x7fddcbc1c1a0 (0@[0:0]-[0:31]) 0 <>
  // String_Quoted 0x7fddcbc07460 (0@[0:0]-[0:31]) [/* url-test1: without quotes */] [delayed] <>
  Import 0x7fddcbc1c230 (0@[1:0]-[1:7]) 0
 @ String_Quoted 0x7fddcbc1c570 (0@[1:8]-[1:12]) [url(a b)] < >
 Comment 0x7fddcbc1c750 (0@[4:0]-[4:28]) 0 <\n\n>
 // String_Quoted 0x7fddcbc07ad0 (0@[4:0]-[4:28]) [/* url-test1: with quotes */] [delayed] <\n\n>
 Import 0x7fddcbc1c690 (0@[5:0]-[5:7]) 0
@ String_Quoted 0x7fddcbc1ca10 (0@[5:8]-[5:12]) [url("a\a b")] < >
 Comment 0x7fddcbc1cb60 (0@[8:0]-[8:28]) 0 <\n\n>
 // String_Quoted 0x7fddcbc08050 (0@[8:0]-[8:28]) [/* url-test2: with quotes */] [delayed] <\n\n>
 Import 0x7fddcbc1caa0 (0@[9:0]-[9:7]) 0
@ String_Quoted 0x7fddcbc1cdd0 (0@[9:8]-[9:12]) [url("a   \r\\d \\\r\\\\d\a     c t n \\t \\n \\t \\n \\\\t \\\\n")] < >
 Comment 0x7fddcbc1ceb0 (0@[12:0]-[12:31]) 0 <\n  \n>
 // String_Quoted 0x7fddcbc08570 (0@[12:0]-[12:31]) [/* url-test2: without quotes */] [delayed] <\n  \n>
 Import 0x7fddcbc1cf40 (0@[13:0]-[13:7]) 0
@ String_Quoted 0x7fddcbc1d2e0 (0@[13:8]-[13:12]) [url(a \d \\d \\\d \\\\d c \t \n \\t \\n \\\t \\\n \\\\t \\\\n)] < >
####################################################################
```

Windows:

```
####################################################################
Block 00B236C0 (0@[0:0]-[15:1]) [root] 0
 Comment 00B1F080 (0@[0:0]-[0:31]) 0 <>
 // String_Quoted 00AEA178 (0@[0:0]-[0:31]) [/* url-test1: without quotes */] [delayed] <>
 Import 00B06B68 (0@[1:0]-[1:7]) 0
@ String_Quoted 00B06DA8 (0@[1:8]-[1:12]) [url(a b)] < >
 Comment 00B1F0F8 (0@[4:0]-[4:28]) 0 <\r\n\r\n>
 // String_Quoted 00AEC078 (0@[4:0]-[4:28]) [/* url-test1: with quotes */] [delayed] <\r\n\r\n>
 Import 00B06FE8 (0@[5:0]-[5:7]) 0
@ String_Quoted 00B07228 (0@[5:8]-[5:12]) [url("a\r\a b")] < >
 Comment 00B1EEA0 (0@[8:0]-[8:28]) 0 <\r\n\r\n>
 // String_Quoted 00AEC1A8 (0@[8:0]-[8:28]) [/* url-test2: with quotes */] [delayed] <\r\n\r\n>
 Import 00B273B0 (0@[9:0]-[9:7]) 0
@ String_Quoted 00B27680 (0@[9:8]-[9:12]) [url("a   \r\\d \\\r\\\\d\r\a     c t n \\t \\n \\t \\n \\\\t \\\\n")] < >
 Comment 00B1F170 (0@[12:0]-[12:31]) 0 <\r\n  \r\n>
 // String_Quoted 00B06808 (0@[12:0]-[12:31]) [/* url-test2: without quotes */] [delayed] <\r\n  \r\n>
 Import 00B28400 (0@[13:0]-[13:7]) 0
@ String_Quoted 00B27F80 (0@[13:8]-[13:12]) [url(a \d \\d \\\d \\\\d c \t \n \\t \\n \\\t \\\n \\\\t \\\\n)] < >
####################################################################
```

Line 10 indicates

Mac: `@ String_Quoted 0x7fddcbc1ca10 (0@[5:8]-[5:12]) [url("a\a b")] < >`
Win: `@ String_Quoted 00B07228 (0@[5:8]-[5:12]) [url("a\r\a b")] < >`

You are right, this issue was present before your fix: 2ecaa6667c837715b705c5298d18f849cfb54735. Note that before 15913abcab246cca2fc49e9772789a6f8ce24e51, my sample was throwing SIGSEGV on both OSes, so I tested just one case: `url-test1: with quotes`. :smile:

**Comment by xzyfer** [comment]
Interesting. Looks like it's worth creating an issue for. Not exactly sure how I'll debug it though haha

You can `debug_ast(root)` after each previous assignment of `root` to see if it's the [parser](https://github.com/sass/libsass/blob/master/src/context.cpp#L626), or [eval](https://github.com/sass/libsass/blob/master/src/context.cpp#L642) step that introduces the `\r`.
